### ISSUE TITLE
Cannot override k3s default kubelet-args via user-provided kubelet config file

### ISSUE THREAD

**Issue by fmunteanu** [k3s-io/k3s#10399]
**Environmental Info:**
K3s Version: 

```
k3s version v1.29.4+k3s1 (94e29e2e)
go version go1.21.9
```

Node(s) CPU architecture, OS, and Version: 

```
Linux apollo 6.8.0-1005-raspi #5-Ubuntu SMP PREEMPT_DYNAMIC Wed May 22 15:01:35 UTC 2024 aarch64 aarch64 aarch64 GNU/Linux
```

Cluster Configuration:

- 3 control planes
- 5 workers

**Describe the bug:**

While using:

```
kubelet-arg:
  - config=/etc/rancher/k3s/kubelet.yaml
```

With `/etc/rancher/k3s/kubelet.yaml` contents:

```
apiVersion: kubelet.config.k8s.io/v1beta1
kind: KubeletConfiguration
evictionHard:
  memory.available: "500Mi"
  nodefs.available: "10%"
  nodefs.inodesFree: "5%"
evictionSoft:
  memory.available: "1Gi"
  nodefs.available: "30%"
  nodefs.inodesFree: "15%"
evictionSoftGracePeriod:
  memory.available: "30s"
  nodefs.available: "30s"
  nodefs.inodesFree: "30s"
evictionMaxPodGracePeriod: 60
systemReserved:
  cpu: "1"
  ephemeral-storage: "5Gi"
  memory: "1Gi"
```

`evictionHard` arg/values are ignored, while other settings are properly reflected, when I run:

```
$ curl -X GET http://127.0.0.1:8001/api/v1/nodes/apollo/proxy/configz -s | jq .kubeletconfig
  "evictionHard": {
    "imagefs.available": "5%",
    "nodefs.available": "5%"
  },
  "evictionSoft": {
    "memory.available": "1Gi",
    "nodefs.available": "30%",
    "nodefs.inodesFree": "15%"
  },
  "evictionSoftGracePeriod": {
    "memory.available": "30s",
    "nodefs.available": "30s",
    "nodefs.inodesFree": "30s"
  },
  "evictionMaxPodGracePeriod": 60
```

**Steps To Reproduce:**

Current `config.yaml`:

```
# cat /etc/rancher/k3s/config.yaml
bind-address: 192.168.4.2
cluster-dns: 10.43.0.10
cluster-domain: cluster.local
disable:
  - local-storage
  - servicelb
  - traefik
disable-cloud-controller: true
disable-kube-proxy: true
disable-network-policy: true
embedded-registry: true
etcd-expose-metrics: true
flannel-backend: none
kubelet-arg:
  - config=/etc/rancher/k3s/kubelet.yaml
node-taint:
  - node.cilium.io/agent-not-ready:NoExecute
  - node-role.kubernetes.io/control-plane:NoSchedule
server: https://192.168.4.10:6443
tls-san:
  - 192.168.4.10
token: redacted
```

However, if I pass directly the arg/values into `config.yaml`, everything works as expected.

New `config.yaml`:

```
# cat /etc/rancher/k3s/config.yaml
bind-address: 192.168.4.2
cluster-dns: 10.43.0.10
cluster-domain: cluster.local
disable:
  - local-storage
  - servicelb
  - traefik
disable-cloud-controller: true
disable-kube-proxy: true
disable-network-policy: true
embedded-registry: true
etcd-expose-metrics: true
flannel-backend: none
kubelet-arg:
  - eviction-hard=memory.available<500Mi,nodefs.available<10%,nodefs.inodesFree<5%
  - eviction-max-pod-grace-period=60
  - eviction-soft=memory.available<1Gi,nodefs.available<30%,nodefs.inodesFree<15%
  - eviction-soft-grace-period=memory.available=30s,nodefs.available=30s,nodefs.inodesFree=30s
  - system-reserved=cpu=1,ephemeral-storage=5Gi,memory=1Gi
node-taint:
  - node.cilium.io/agent-not-ready:NoExecute
  - node-role.kubernetes.io/control-plane:NoSchedule
server: https://192.168.4.10:6443
tls-san:
  - 192.168.4.10
token: redacted
```

Settings are correct:

```
$ curl -X GET http://127.0.0.1:8001/api/v1/nodes/apollo/proxy/configz -s | jq .kubeletconfig
  "evictionHard": {
    "memory.available": "500Mi",
    "nodefs.available": "10%",
    "nodefs.inodesFree": "5%"
  },
  "evictionSoft": {
    "memory.available": "1Gi",
    "nodefs.available": "30%",
    "nodefs.inodesFree": "15%"
  },
  "evictionSoftGracePeriod": {
    "memory.available": "30s",
    "nodefs.available": "30s",
    "nodefs.inodesFree": "30s"
  },
  "evictionMaxPodGracePeriod": 60
```

Edit: Same issue occurs for `evictionMinimumReclaim`, it only works if you pass the values directly, not through `kubelet.yaml` configuration:

```
evictionMinimumReclaim:
  imagefs.available: "10%"
  imagefs.inodesFree: "5%"
  memory.available: "500Mi"
  nodefs.available: "10%"
  nodefs.inodesFree: "5%"
```

Actual settings:

```
$ curl -X GET http://127.0.0.1:8001/api/v1/nodes/apollo/proxy/configz -s | jq .kubeletconfig
  "evictionMinimumReclaim": {
    "imagefs.available": "10%",
    "nodefs.available": "10%"
  }
```

**Comment by brandond** [comment]
We don't modify any of the code related to the kubelet's parsing of the config file, so it's unclear to me what part of this is a k3s issue. Have you tried turning up the kubelet's verbosity with -v=9 to see if it is failing to parse some bit of your config file?

**Comment by fmunteanu** [comment]
@brandond i have no idea how to set the kubelet verbosity with K3s. From my perspective, the `config` argument is passed through `kubelet-arg`, therefore is directly related to K3s.

**Comment by brandond** [comment]
> turning up the kubelet's verbosity with --v=9

Just add that flag to the k3s args, or `v: 9` if you're using a k3s config.yaml.

**Comment by fmunteanu** [comment]
@brandond I found the root cause of the curent issue and IMO is a quite important one. I'm 90% sure is related to K3s, since I don't have access to a K8s cluster. Let explain the problem, below.

When K3s service is started on a Linux environment, it passes a list of parameters to the binary:

```
# kubectl cordon apollo
# journalctl -u k3s --rotate --vacuum-time=1s
# systemctl restart k3s
# journalctl -u k3s | grep 'msg="Running kubelet' | sed 's| --|\n  --|g'
Jun 25 19:03:01 apollo k3s[22973]: time="2024-06-25T19:03:01-04:00" level=info msg="Running kubelet
  --address=0.0.0.0
  --anonymous-auth=false
  --authentication-token-webhook=true
  --authorization-mode=Webhook
  --cgroup-driver=systemd
  --client-ca-file=/var/lib/rancher/k3s/agent/client-ca.crt
  --cluster-dns=10.43.0.10
  --cluster-domain=cluster.local
  --config=/etc/rancher/k3s/kubelet.yaml
  --container-runtime-endpoint=unix:///run/k3s/containerd/containerd.sock
  --containerd=/run/k3s/containerd/containerd.sock
  --eviction-hard=imagefs.available<5%,nodefs.available<5%
  --eviction-minimum-reclaim=imagefs.available=10%,nodefs.available=10%
  --fail-swap-on=false
  --healthz-bind-address=127.0.0.1
  --hostname-override=apollo
  --kubeconfig=/var/lib/rancher/k3s/agent/kubelet.kubeconfig
  --node-ip=192.168.4.2
  --node-labels=
  --pod-infra-container-image=rancher/mirrored-pause:3.6
  --pod-manifest-path=/var/lib/rancher/k3s/agent/pod-manifests
  --read-only-port=0
  --register-with-taints=node.cilium.io/agent-not-ready:NoExecute,node-role.kubernetes.io/control-plane:NoSchedule
  --resolv-conf=/run/systemd/resolve/resolv.conf
  --serialize-image-pulls=false
  --tls-cert-file=/var/lib/rancher/k3s/agent/serving-kubelet.crt
  --tls-private-key-file=/var/lib/rancher/k3s/agent/serving-kubelet.key"
```

Note the 3 key arguments:

```
  --config=/etc/rancher/k3s/kubelet.yaml
  --eviction-hard=imagefs.available<5%,nodefs.available<5%
  --eviction-minimum-reclaim=imagefs.available=10%,nodefs.available=10%
```

Since the command arguments take precedence of arguments passed through `kubelet.yaml`, it explains why they are ignored. Passing the `--eviction-hard` and `--eviction-minimum-reclaim` at service start is really bad, because it tells to kubelet to ignore the [default settings](https://kubernetes.io/docs/reference/config-api/kubelet-config.v1beta1/#kubelet-config-k8s-io-v1beta1-KubeletConfiguration):

> Default: memory.available: "100Mi" nodefs.available: "10%" nodefs.inodesFree: "5%" imagefs.available: "15%" imagefs.inodesFree: "5%"

Setting only these two arg/values will automatically ignore the `memory.available`, `nodefs.inodesFree` and `imagefs.inodesFree` values:

```
  --eviction-hard=imagefs.available<5%,nodefs.available<5%
```

And can result on a major performance impact, since kubelet ignores the non-defined hard evictions. Say, you have a bad application with memory leak, it will quickly fill the node memory, until kubelet crashes. I need your help or any of the devs to look at K3s source code and see how the `--eviction-hard` and `--eviction-minimum-reclaim` flags are passed. We need to have these two flags disabled and be controlled by end-user only. This way, kubelet can properly set the default garbage collection values, which protects the cluster from catastrophic failures.

I traced this behaviour all the way to K3s 1.22, see https://github.com/k3s-io/k3s/issues/4509#issuecomment-974486948.

**Comment by fmunteanu** [comment]
There two lines should removed:

https://github.com/k3s-io/k3s/blob/master/pkg/daemons/agent/agent_linux.go#L79-L80

**Comment by fmunteanu** [comment]
@brandond I created a [PR](https://github.com/k3s-io/k3s/pull/10404) with the proposed fix.

**Comment by brandond** [comment]
Related: https://github.com/k3s-io/k3s/issues/2116

I would consider our support for kubelet configuration files somewhat limited until we address that issue, due to how the kubelet handles configuration passed via both config AND cli.

**Comment by fmunteanu** [comment]
That makes sense @brandond, any idea when this is planned to be pushed to stable release?

**Comment by brandond** [comment]
1. We are not currently planning on changing the default GC behavior, as it has been working for our users/customers for the last 6 years. We will improve the ability to alter our defaults, but that is as much as I am willing to commit to at the moment.
2. We don't currently have a target date for switching to kubelet config instead of CLI args. Upstream shows no signs of actually removing CLI-based configuration, despite it being deprecated (and deprecated for only the kubelet, not any other component) for several years.

**Comment by fmunteanu** [comment]
Thank you @brandond for taking the time to discuss it with the devs, internally. At the end of the line, if end-user knows where is the issue, he/she knows how to fix it. I really appreciate you spending time on this.

Like you mentioned earlier, the forced kubelet settings can be fixed by passing new args to override the defaults you have hardcoded. I was simply trying to reinforce the defaults presented in K8s.

I’m going to close this issue because I consider what we have now implemented is satisfactory. I’m glad we sorted everything, it will be useful for other users. I started looking into kubelet settings when I saw a cluster kubelet crashing because it was out of memory.
